Add metrics uploader client

IrinaShkviro opened the pull request:
Add MetricsUploader subproject to send log events from Orbit. For
communication between Orbit and metrics_uploader_client.dll
OrbitLogEvent proto message is used.

Currently only ORBIT_INITIALIZED and ORBIT_CAPTURE_DURATION
log message are available.

The metrics will be sent only if metrics_uploader_client.dll is available
and --enable_metrics flag is set.

Bug: http://b/170389366

Test: start Orbit, check that the ORBIT_INITIALIZED and
ORBIT_CAPTURE_DURATION log messages appears on the server.

beckerhe commented on MetricsUploader/include/MetricsUploader/MetricsUploader.h:
```diff
@@ -0,0 +1,42 @@
+  void SendLogEvent(orbit_client_protos::OrbitLogEvent_LogEventType log_event_type);
+
+ private:
+  inline static bool is_init_{false};
```
We should make that class thread-safe by default which means, `is_init_` should have a mutex.

IrinaShkviro replied:
Thanks, I've added Mutex for `is_init_`.

beckerhe commented on MetricsUploader/MetricsUploaderWindows.cpp:
```diff
@@ -0,0 +1,68 @@
+constexpr const char* kStartUploaderClientFunctionName = "StartUploaderClient";
+
+MetricsUploader::MetricsUploader() {
+  CHECK(!is_init_);
```
I think this can be made more explicit for the user of this class.

You can make the constructor private and add a factory function which returns a `ErrorMessageOr<MetricsUploader>`. So everybody has to go through the factory function and check if they actually got a MetricsUploader object.

A singleton might be another solution.

IrinaShkviro replied:
Added `static CreateMetricsUploader()` function and made constructor private.

beckerhe commented on OrbitClientProtos/orbit_log_event.proto:
```diff
@@ -0,0 +1,15 @@
+    ORBIT_INITIALIZED = 1;
+  }
+  LogEventType log_event_type = 1;
+}
```
That's the proto which have to be kept in sync with the library, isn't it? Could you add a comment explaining that to avoid folks accidentally changing that without making the some change in the library?

IrinaShkviro replied:
Yes, right, added a comment about it, thank you!

beckerhe commented on MetricsUploader/include/MetricsUploader/MetricsUploader.h:
```diff
@@ -0,0 +1,42 @@
+#define ORBIT_METRICS_UPLOADER_METRICS_UPLOADER_H_
+
+#ifdef _WIN32
+#include <windows.h>
```
It's only about HINSTANCE, isn't it? So `#include <WinDef.h>` should be enough.

IrinaShkviro replied:
Including only `<WinDef.h>` cause `fatal error C1189: #error:  "No Target Architecture"` for me.

beckerhe commented on MetricsUploader/include/MetricsUploader/MetricsUploader.h:
```diff
@@ -0,0 +1,42 @@
+ private:
+  inline static bool is_init_{false};
+#ifdef _WIN32
+  HINSTANCE metrics_uploader_client_dll_;
```
Nit: AFAIK the return type of LoadLibrary is HMODULE (which happens to be the same as HINSTANCE - but ideally we don't assume that.)

IrinaShkviro replied:
Right, fixed it, than you!

beckerhe commented on OrbitClientProtos/CMakeLists.txt:
```diff
@@ -8,6 +8,7 @@ project(OrbitClientProtos)
 
 add_library(OrbitClientProtos STATIC)
 protobuf_generate(TARGET OrbitClientProtos PROTOS capture_data.proto
+                                                  orbit_log_event.proto
```
I think that proto should live in `MetricsUploader` instead. It's tied to the class in there, so there is no use case without the class.

IrinaShkviro replied:
I've moved it to the `MetricsUploader/proto` folder.

beckerhe commented on MetricsUploader/include/MetricsUploader/MetricsUploader.h:
```diff
@@ -0,0 +1,42 @@
+  inline static bool is_init_{false};
+#ifdef _WIN32
+  HINSTANCE metrics_uploader_client_dll_;
+  void (*send_log_event_addr_)(const uint8_t*, int);
```
Can you initialize this to `nullptr`?

IrinaShkviro replied:
Done, thanks.

beckerhe commented on MetricsUploader/MetricsUploaderWindows.cpp:
```diff
@@ -0,0 +1,68 @@
+}
+
+void MetricsUploader::SendLogEvent(OrbitLogEvent_LogEventType log_event_type) {
+  if (is_init_) {
```
Nit: `if (send_log_event_addr_ != nullptr)` instead

That's basically the same condition but easier to understand for someone just looking at this function.

IrinaShkviro replied:
Changed the condition to `if (send_log_event_addr_ != nullptr)`.

beckerhe commented on MetricsUploader/MetricsUploaderWindows.cpp:
```diff
@@ -0,0 +1,68 @@
+
+MetricsUploader::~MetricsUploader() {
+  if (is_init_) {
+    FreeLibrary(metrics_uploader_client_dll_);
```
I guess you should reset is_init_ here.

IrinaShkviro replied:
Added `Reset()` function for this, thank you!

beckerhe reviewed (commented):
Hi Irina,
sorry for being so late. I totally missed when that moved from Draft to Ready.

I added some comments, but overall looks good.

But let's quickly talk about tests. :innocent: 

I think that code needs unit tests. And I gave some thought on how to do that.

1. I wouldn't make the library file name fixed. In that case you can load a different DLL in a test.
2. I would add a DLL which implements the same functions as the uploader library. Then you can load that DLL in a test case and see if it behaves correctly.

beckerhe commented on MetricsUploader/include/MetricsUploader/MetricsUploader.h:
```diff
@@ -0,0 +1,44 @@
+
+namespace orbit_metrics_uploader {
+
+/**
+ * This class is used for sending log events from Orbit. It works only on Windows and only if
+ * metrics_uploader_client.dll is available. The types of logs that could be send are defined in
+ * orbit_client_protos::OrbitLogEvent::LogEventType.
+ */
```
The text also needs to describe the pre-conditions. In this case: The class can only be instantiated once. But like I said in other comments, ideally we make that clear on the API level and don't rely on comments or CHECKS.

IrinaShkviro replied:
I've changed the description, ptal

IrinaShkviro commented on MetricsUploader/MetricsUploaderLinux.cpp:
```diff
@@ -0,0 +1,20 @@
+
+MetricsUploader::MetricsUploader() {}
+
+MetricsUploader::~MetricsUploader() {}
```
Done, thanks

IrinaShkviro commented on MetricsUploader/MetricsUploaderWindows.cpp:
```diff
@@ -0,0 +1,76 @@
+    log_event.set_event_duration_milliseconds(event_duration_ms);
+
+    int message_size = log_event.ByteSize();
+    uint8_t* buffer = new uint8_t[message_size];
```
I've added std::vector for the buffer, thanks.

IrinaShkviro commented on MetricsUploader/MetricsUploaderWindows.cpp:
```diff
@@ -0,0 +1,76 @@
+    send_log_event_addr_(buffer, message_size);
+    delete[] buffer;
+  }
+}
```
Right, I've moved it to the MetricsUploader.cpp.

IrinaShkviro commented:
> Hi Irina,
> sorry for being so late. I totally missed when that moved from Draft to Ready.
> 
> I added some comments, but overall looks good.
> 
> But let's quickly talk about tests. 😇
> 
> I think that code needs unit tests. And I gave some thought on how to do that.
> 
> 1. I wouldn't make the library file name fixed. In that case you can load a different DLL in a test.
> 2. I would add a DLL which implements the same functions as the uploader library. Then you can load that DLL in a test case and see if it behaves correctly.

Thanks a lot for the review Henning! I've added several tests and also addressed most of the comments. Could you, please take another look?

ronaldfw commented on MetricsUploader/proto/orbit_log_event.proto:
```diff
@@ -0,0 +1,22 @@
+  LogEventType log_event_type = 1;
+
+  string orbit_version = 2;
+  int64 event_duration_milliseconds = 3;
```
Do all events necessarily have a duration? And maybe we want a timestamp of when an event occurred? That would allow computations of durations after the fact. E.g. types ORBIT_CAPTURE_STARTED and ORBIT_CAPTURE_FINISHED with a timestamp each. Or does that cause other problems in the analysis of the data?

IrinaShkviro replied:
Not all the events have duration, this field is optional. There were 2 options that I've considered: to send event with duration or to send 2 events (start and end) and use sql query to calculate the duration. From the metrics uploader client perspective they look pretty similar, but I guess that having duration field would make queries easier (for example, when we want to get mean/max capture length for the specific time period). If we want to use ORBIT_CAPTURE_STARTED and ORBIT_CAPTURE_FINISHED only to calculate the duration I think it would be better just to send duration event, or am I missed something?

beckerhe commented on MetricsUploader/MetricsUploader.cpp:
```diff
@@ -0,0 +1,69 @@
+  if (is_init_) {
+    return ErrorMessage("MetricsUploader is already created");
+  }
+  MetricsUploader* metrics_uploader = new MetricsUploader(client_name);
```
You could use std::make_unique here instead of `new`.

IrinaShkviro replied:
I'm not sure I can do it due to private constructor: https://abseil.io/tips/134

beckerhe replied:
Ah, yes you're right. In this case I'd would recommend to move absl::WrapUnique to line 43. This way you still have a unique_ptr the whole.

BTW, you can also write `const auto metrics_uploader = std::unique_ptr{new MetricsUploader(client_name)};`. AFAIK `WrapUnique` also just calls the constructor of unique_ptr.

IrinaShkviro replied:
Done

ronaldfw reviewed (commented):
I don't have much context about this at this point, but would like to discuss the log event definitions. Not sure if my comments make sense, but would love to understand our options better.

beckerhe commented on MetricsUploader/include/MetricsUploader/MetricsUploader.h:
```diff
@@ -0,0 +1,76 @@
+
+ private:
+  // Load metrics uploader client library if available and starts metrics uploader client.
+  MetricsUploader(const std::string_view client_name = kMetricsUploaderClientDllName);
```
`explicit`

IrinaShkviro replied:
fixed, thanks

beckerhe commented on MetricsUploader/include/MetricsUploader/MetricsUploader.h:
```diff
@@ -0,0 +1,76 @@
+  bool SendLogEvent(OrbitLogEvent_LogEventType log_event_type);
+
+ private:
+  // Load metrics uploader client library if available and starts metrics uploader client.
+  MetricsUploader(const std::string_view client_name = kMetricsUploaderClientDllName);
+
+  void Reset();
```
If I understand everything correctly: The user should call `CreateMetricsUploader` and will only get a result if no instance has been created before. So that's good.

If no instance has been created yet, the CreateMetricsUploader function will create a new `MetricsUploader` instance. But the construction of that instance can fail which is checked, by `is_init_`.

What about moving all the code from the constructor to `CreateMetricsUploader`? This would simplify the design since `Reset won't be needed anymore and `is_init` will only be changed in one place - in `CreateMetricsUploader`

IrinaShkviro replied:
I reorganized this part of the code, could you please take another look?

beckerhe commented on MetricsUploader/include/MetricsUploader/MetricsUploader.h:
```diff
@@ -0,0 +1,76 @@
+  void Reset();
+  std::string GetErrorMessage(Result error);
+
+  inline static bool is_init_ = false;
+  static absl::Mutex init_mutex_;
```
Personal preference: If `is_init_` was inline, I would also make `init_mutex_` inline. Then you don't need the definition in the cpp file.

IrinaShkviro replied:
I've changed this code: now I have a non-static mutex which id used for send_log_event, and check if the library was loaded earlier with GetModuleHandleExA function.

beckerhe commented on OrbitGl/App.cpp:
```diff
@@ -126,8 +127,9 @@ PresetLoadState GetPresetLoadStateForProcess(
 bool DoZoom = false;
 
-OrbitApp::OrbitApp(MainThreadExecutor* main_thread_executor)
-    : main_thread_executor_(main_thread_executor) {
+OrbitApp::OrbitApp(MainThreadExecutor* main_thread_executor,
+                   orbit_metrics_uploader::MetricsUploader* metrics_uploader)
+    : main_thread_executor_(main_thread_executor), metrics_uploader_(metrics_uploader) {
```
Could you put all the integration changes in OrbitGl and OrbitQt into a separate commit?

IrinaShkviro replied:
Done

IrinaShkviro commented on MetricsUploader/proto/orbit_log_event.proto:
```diff
@@ -0,0 +1,22 @@
+  enum LogEventType {
+    UNKNOWN_EVENT_TYPE = 0;
+    ORBIT_INITIALIZED = 1;
+    ORBIT_CAPTURE_DURATION = 2;
```
Here I've used the similar to others teams approach (to have a separate log type for every event). Having a comment describing an event could be another option, but what's advantage of it?

ronaldfw replied:
It would potentially give more flexibility as you don't have to change the LogEventType definition if you introduce a new event somewhere. But if this is common practice, it seems fine to me.

beckerhe commented on MetricsUploader/include/MetricsUploader/MetricsUploader.h:
```diff
@@ -53,7 +53,7 @@ class MetricsUploader {
   // Send log events to the server using metrics_uploader.
   // Returns true on success and false otherwise.
-  bool SendLogEvent(OrbitLogEvent_LogEventType log_event_type);
+  bool SendLogEvent(OrbitLogEvent_LogEventType log_event_type, int64_t event_duration_ms = 0);
```
What about using `std::chrono::milliseconds` instead of  `int64_t`?

IrinaShkviro replied:
Changed to `std::chrono::milliseconds`, thanks!

beckerhe commented on OrbitGl/App.cpp:
```diff
@@ -871,6 +871,17 @@ void OrbitApp::StopCapture() {
   }
 
+  if (metrics_uploader_ != nullptr) {
+    thread_pool_->Schedule([this] {
```
What is the reason for doing this asynchronously? Does it take a long time to send a log event?

I would expect a couple of milliseconds max. That shouldn't be a problem for the user since this is just called once.

IrinaShkviro replied:
Remove async call for now, will modify the MetricsUploader itself to make async calls if needed in future

beckerhe commented on MetricsUploader/TestClients/SendEventClient.cpp:
```diff
@@ -0,0 +1,17 @@
+
+#include "MetricsUploader/Result.h"
+
+#define DllExport __declspec(dllexport)
```
Currently there is no reason for having this macro. It's usually there because you surround it by `#ifdef _WIN32` and `#endif`.

So I'd either add the `ifdef` or remove it.

IrinaShkviro replied:
Removed `#define ` here

beckerhe commented on MetricsUploader/TestClients/StartAndSendEventClient.cpp:
```diff
@@ -0,0 +1,30 @@
+using orbit_metrics_uploader::Result;
+
+extern "C" {
+DllExport enum Result StartUploaderClient() { return Result::kNoError; }
```
Is `enum Result` correct here? Shouldn't this be `int`? What does the header that is generated from the go-code show? Is there an enum?

IrinaShkviro replied:
I put the same enum in the preamble, so generated header have the `enum Result` declaration. The signature of the functions are:
```
extern enum Result StartUploaderClient();
extern enum Result SendOrbitLogEvent(uint8_t* serializedProto, int length);
```

beckerhe commented on MetricsUploader/MetricsUploaderWindowsTest.cpp:
```diff
@@ -0,0 +1,54 @@
+  EXPECT_EQ(result, true);
+}
+
+TEST(MetricsUploader, SecondMetricsUploader) {
```
What do you mean by `SecondMetricsUploader`?

IrinaShkviro replied:
In this test I want to check that it's not possible to create MetricsUploader for the second time. Renamed this test

beckerhe reviewed (commented):
Very cool Irina. This looks really good now. I added some questions and comments.

beckerhe commented on MetricsUploader/MetricsUploaderWindows.cpp:
```diff
@@ -0,0 +1,75 @@
+MetricsUploader::~MetricsUploader() {
+  absl::MutexLock lock(&send_log_event_mutex_);
+  send_log_event_addr_ = nullptr;
+  HMODULE metrics_uploader_client_dll;
+  if (GetModuleHandleExA(GET_MODULE_HANDLE_EX_FLAG_UNCHANGED_REFCOUNT, client_name_.c_str(),
+                         &metrics_uploader_client_dll) != 0) {
+    FreeLibrary(metrics_uploader_client_dll);
+  }
```
Could you make `metrics_uploader_client_dll` a member variable of `MetricsUploader` and then just call `FreeLibrary`?

IrinaShkviro replied:
Done

beckerhe commented on MetricsUploader/MetricsUploaderWindows.cpp:
```diff
@@ -0,0 +1,75 @@
+
+MetricsUploader::~MetricsUploader() {
+  absl::MutexLock lock(&send_log_event_mutex_);
+  send_log_event_addr_ = nullptr;
```
Setting this to `nullptr` is not necessary, since the object gets destructed anyway.

IrinaShkviro replied:
Right, thank you

beckerhe commented on MetricsUploader/MetricsUploaderWindows.cpp:
```diff
@@ -0,0 +1,75 @@
+    }
+
+    auto metrics_uploader = std::unique_ptr<MetricsUploader>{new MetricsUploader(client_name)};
+    metrics_uploader->send_log_event_addr_ = (Result(*)(const uint8_t*, int))GetProcAddress(
```
Instead of assigning `send_log_event_addr_`, could you make this a constructor argument?

beckerhe replied:
`HMODULE metrics_uploader_client_dll` also needs to be a constructor argument (see comment above).

IrinaShkviro replied:
Made both variables constructor arguments (for Windows)

beckerhe commented on MetricsUploader/include/MetricsUploader/MetricsUploader.h:
```diff
@@ -0,0 +1,71 @@
+  bool SendLogEvent(OrbitLogEvent_LogEventType log_event_type);
+
+ private:
+  explicit MetricsUploader(std::string_view client_name);
```
Since `client_name` is a sink variable, you should use `std::string` here and move it into place.

IrinaShkviro replied:
Done, thanks

beckerhe commented on MetricsUploader/include/MetricsUploader/MetricsUploader.h:
```diff
@@ -0,0 +1,71 @@
+  ~MetricsUploader();
+
+  MetricsUploader(const MetricsUploader& other) = delete;
+  MetricsUploader(MetricsUploader&& other) = delete;
```
With the new way of handling `FreeLibrary` a move constructor/assignment operator can be easily defined:

```
MetricsUploader(MetricsUploader&& other) : 
  send_log_event_addr_{other.send_log_event_addr_},
  client_name{std::move(client_name)}
{
  other.send_log_event_addr_ = nullptr;
}
```

The move assignment-operator is roughly the same, but needs to check for self assignment:
```
MetricsUploader& MetricsUploader::operator= (MetricsUploader&& other) {
  if(&other == this) return *this;

  send_log_event_addr_ = other.send_log_event_addr_;
  other.send_log_event_addr_ = nullptr;

  client_name = std::move(other.client_name);
  return *this;
}
```

And the destructor may only call `FreeLibrary` when `send_log_event_addr_` is `nullptr`:

```
MetricsUploader::~MetricsUploader() {
  if (send_log_event_addr_ == nullptr)
    return;

  // destructor work....
}
```

beckerhe replied:
You would also need to move `metrics_uploader_client_dll_` (which I suggested to be a member variable in a comment above)

IrinaShkviro replied:
Added move constructor/assignment for the MetricsUploader for Windows

beckerhe commented on MetricsUploader/include/MetricsUploader/MetricsUploader.h:
```diff
@@ -0,0 +1,71 @@
+  // Create a MetricsUploader instance, load metrics uploader client library if available and starts
+  // metrics uploader client when called first time. Return the instance if there are no errors and
+  // ErrorMessage otherwise.
+  [[nodiscard]] static ErrorMessageOr<std::unique_ptr<MetricsUploader>> CreateMetricsUploader(
+      std::string_view client_name = kMetricsUploaderClientDllName);
```
When a move constructor is defined (see instructions below) you could return a `ErrorMessageOr<MetricsUploader>` and avoid the heap allocation. To me this is a better design, but I don't feel particularly strong about this.

IrinaShkviro replied:
Agree, removed `unique_ptr` here.

beckerhe commented on MetricsUploader/MetricsUploaderWindows.cpp:
```diff
@@ -0,0 +1,75 @@
+constexpr const char* kStartUploaderClientFunctionName = "StartUploaderClient";
+
+MetricsUploader::~MetricsUploader() {
+  absl::MutexLock lock(&send_log_event_mutex_);
```
You can remove this lock. A destructor can't be thread-safe anyway, since it can only be called once.

IrinaShkviro replied:
Done, thank you!

beckerhe commented on MetricsUploader/include/MetricsUploader/MetricsUploader.h:
```diff
@@ -0,0 +1,71 @@
+  absl::Mutex send_log_event_mutex_;
+  Result (*send_log_event_addr_)(const uint8_t*, int) = nullptr;
+
+  const std::string client_name_;
```
When you define move constructors (like suggested above) this should be non-const. Otherwise you can't move the string to another instance.

IrinaShkviro replied:
Done

beckerhe commented on MetricsUploader/MetricsUploaderWindows.cpp:
```diff
@@ -0,0 +1,75 @@
+
+    return metrics_uploader;
+  }
+  std::string error_message = "Metrcs uploader client library is not found";
```
Nit: Metrics

IrinaShkviro replied:
fixed, thanks

beckerhe commented on MetricsUploader/MetricsUploaderWindows.cpp:
```diff
@@ -0,0 +1,75 @@
+      std::string error_message =
+          absl::StrFormat("%s function not found", kStartUploaderClientFunctionName);
+      ERROR("%s", error_message);
+      FreeLibrary(metrics_uploader_client_dll);
```
You probably should do this in the other error branches below as well, don't you?

beckerhe commented on MetricsUploader/MetricsUploaderWindows.cpp:
```diff
@@ -0,0 +1,75 @@
+
+    Result (*start_uploader_client_addr_)();
+    start_uploader_client_addr_ =
+        (Result(*)())GetProcAddress(metrics_uploader_client_dll, kStartUploaderClientFunctionName);
```
Can you replace the c-style cast by static_cast?

IrinaShkviro replied:
I've added `reinterpret_cast` here

beckerhe commented on MetricsUploader/MetricsUploaderWindows.cpp:
```diff
@@ -0,0 +1,75 @@
+  if (NULL != metrics_uploader_client_dll) {
+    LOG("Metrcs uploader client library is loaded");
+
+    Result (*start_uploader_client_addr_)();
+    start_uploader_client_addr_ =
```
When using `auto` you could avoid declaring the variable first

IrinaShkviro replied:
Right, thanks

beckerhe commented on MetricsUploader/MetricsUploaderWindows.cpp:
```diff
@@ -0,0 +1,75 @@
+    return ErrorMessage("MetricsUploader is already created");
+  }
+
+  metrics_uploader_client_dll = LoadLibraryA(client_name.data());
```
You can't use `.data()`. `data()` is not null-terminated! You need to use `.c_str()` which is only available in `std::string`, so you have to replace the parameter type.

IrinaShkviro replied:
thank you, fixed!

beckerhe commented on MetricsUploader/MetricsUploaderWindows.cpp:
```diff
@@ -0,0 +1,75 @@
+ErrorMessageOr<std::unique_ptr<MetricsUploader>> MetricsUploader::CreateMetricsUploader(
+    std::string_view client_name) {
+  HMODULE metrics_uploader_client_dll;
+  if (GetModuleHandleExA(GET_MODULE_HANDLE_EX_FLAG_UNCHANGED_REFCOUNT, client_name.data(),
```
You can't use `.data()`. `data()` is not null-terminated! You need to use `.c_str()` which is only available in `std::string`, so you have to replace the parameter type.

IrinaShkviro replied:
Done

beckerhe commented on MetricsUploader/MetricsUploaderLinux.cpp:
```diff
@@ -0,0 +1,16 @@
+
+MetricsUploader::~MetricsUploader() = default;
+
+ErrorMessageOr<std::unique_ptr<MetricsUploader>> MetricsUploader::CreateMetricsUploader([
+    [maybe_unused]] std::string_view client_name) {
```
The line break looks weird. Seems to be a clang-format problem. I'd suggest to not specify the argument name (client_name). Then the attribute is not needed in the first place.

IrinaShkviro replied:
Yes, strange for me as well. Removed argument name and it looks better now

beckerhe reviewed (commented):
Hi Irina,
I added some more comments. I'm sorry I hope you're not already fed up with my comments. :see_no_evil: 

Thank you for the changes!

Henning

IrinaShkviro commented:
> Hi Irina,
> I added some more comments. I'm sorry I hope you're not already fed up with my comments. 🙈
> 
> Thank you for the changes!
> 
> Henning

Thanks a lot for the review Henning! I've addressed comments and I think it looks much better now =)

beckerhe commented on MetricsUploader/include/MetricsUploader/MetricsUploader.h:
```diff
@@ -0,0 +1,78 @@
+//
+// auto metrics_uploader = MetricsUploader::CreateMetricsUploader();
+// if (metrics_uploader.has_value()) {
+//   metrics_uploader.value()->SendLogEvent(...);
```
Since `CreateMetricsUploader` now returns a value, it's either `metrics_uploader.value().SendLogEvent` or `metrics_uploader->SendLogEvent`.

IrinaShkviro replied:
Thanks a lot! updated the comment

beckerhe reviewed (approved):
Great! LGTM now! Thank you, Irina.

I added a minor comment regarding the example in the comments.
